Update on maiana-wifi

marcelrv opened the pull request:
I've made quite some additions to your work. 

* Added OTA update
* Added bi-directional communication from TCP (meaning you can also configure from Openplotter etc)
* Added 'memory' of the last messages from all ships seen in the last 10  minutes and send it on connect
* Changed the port settings to allow multiple communication methods at the same time
* Added websocket as commuinication
* Added a webpage with GPS details
* Split the configuration from the info page
* Added table with AIS information from surrounding ships
* Added live view of NMEA messages
* Added the possibility to set/load the wifi details from serial

and prob several more smaller changes...

marcelrv committed:
major update

* Add NMEA forwarding functionality for multiple types instead of single
* Add websocket support
* Add NMEA0183 messages window
* Load and save wifi settings and protocol settings
* fake AIS data source for testing
* Add signal strenght
* Many smaller edits

marcelrv committed:
Allow to configure wifi via serial

marcelrv committed:
test loading via web

marcelrv committed:
Fix CRC check and add wifi details

marcelrv committed:
update demo messages

marcelrv committed:
Add GPS dashboard

marcelrv committed:
adjust style gps dashboard

marcelrv committed:
readme update

marcelrv committed:
send all recent ais messages on connect

marcelrv committed:
Add GPS dashboard

marcelrv committed:
gps dashboard adjustable size

marcelrv committed:
fix satellite view

marcelrv committed:
Prepare for online flashing of firmware

marcelrv committed:
adds resending of ais messages to ws & fakes messages from file

marcelrv committed:
Modify UI splitting into separate config page and add AIS table

marcelrv committed:
Add main page info and rearrange infobox

marcelrv committed:
Merge ais & index page.

marcelrv committed:
update readme

marcelrv committed:
update readme

marcelrv committed:
add firmware files

marcelrv committed:
Readme update

marcelrv committed:
return accidental removed controls for NMEA message window

muebau commented:
Wow that is impressive. Thank you a lot.
I started going through all the commits. This takes some time as the changed formatting brings a lot of changes and make it hard to find the actual changes.

muebau commented on data/dashboard.html:
```diff
@@ -0,0 +1,560 @@
+        <meta name="viewport" content="initial-scale=1.0">
+        <meta charset="utf-8">
+
+        <link href="https://fonts.googleapis.com/css?family=Oswald" rel="stylesheet">
```
I think there should be no 'online' artifacts.

This could be changed to:
```css
/* oswald-regular - latin */
@font-face {
  font-display: swap; /* Check https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display for other options. */
  font-family: 'Oswald';
  font-style: normal;
  font-weight: 400;
  src: url('/fonts/oswald-v53-latin-regular.woff2') format('woff2'); /* Chrome 36+, Opera 23+, Firefox 39+, Safari 12+, iOS 10+ */
}
```
font file:
https://gwfh.mranftl.com/api/fonts/oswald?download=zip&subsets=latin&variants=regular&formats=woff2

source https://gwfh.mranftl.com/fonts/oswald?subsets=latin

marcelrv replied:
fully agree. 
Need to check if the font file still fits the size of the SPIFF. (I can make the demo/test ais file smaller otherwise)
Alternatively I use the same font you used in in the other pages, then it just needs resizing here &there. (the font came from the GPS example I used)

muebau commented on data/index.html:
```diff
@@ -3,343 +3,352 @@
     </style>
     <link rel="icon" type="image/png"
         href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAACqklEQVQ4T42SbUhTURjHn3Pu3VU3IUq02RKlog9qhGH4oYxK6ItUVKaohB96+1ZBL4ShrtIsKzT8kC9IQe+7YJgVkoWoqYGv2Nzm2DQkndNyq5b3lu6ezr2ytTU/dL6d55z/7+H5P38E/3GIIWUfYMgBQoywwNSjvP4vPhkK1J8oLIvFwjyuqSybDKyTJ8lxwHIJFLANEDoKIBWgw0Pd8h8/4HRx+XlBEK/TEq2RUYRQ5dx4YgPPZ3sJv6WS/twFEqkGaaEbGO4FiMIOdMTsUAC9w9Z1zW/eWqadTtU/E/URFuXXl+utxLBZB5h ...
+    <!--    <script type="module"  src="js/AisParser.js"></script>
```
still needed?

muebau commented on docu/flash copy.html:
```diff
@@ -0,0 +1,52 @@
+4. Connect to the Maiana wifi and configure the wifi network to connect your device to.
+
+<p class="button-row" align="center">
+<esp-web-install-button manifest="https://raw.githubusercontent.com/marcelrv/MAIANA-WiFi/webloader/docu/manifest.json">
```
Could this be made with relative (to the repo) paths?

muebau commented on docu/flash copy.html:
Does it work yet?

marcelrv replied:
yes & no... consider it fully experimental.
I sucessfully flashed my device with it, but running the html file local.
I have yet to find out / test how to make the github page render html
https://docs.github.com/en/pages/getting-started-with-github-pages/creating-a-github-pages-site

it was orginally to be part of separate branch, but than I screwed it up  and added several unrelated commit to it and it became my main branch...
But ultimately, I think it should be easy to flash, and if the whole thing can be flashed in one go it is way easier.
In my boat the esp is not easy to access, hence I really wanted to have OTA, but the laptop I have there also does not have all the development/vscode installed, so the install needed to be easy without need to install the flashing programs.

muebau replied:
Well I love the idea. All big projects (Tasmota, WLED, ect.) offer it. I think this should be part of this project as it might be sensible to expect a very limited environment on a ship. It would be even better if it would work on Android (which it doesn't).

muebau commented on docu/flash.md:
This is work in progress right?

muebau commented on docu/firmware.bin:
I will have a look in 'automatic builds'. I think this should be based on 'releases'.

muebau commented on docu/manifest.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "Maiana",
```
MAIANA would be Peters choice of the name.

muebau commented on docu/manifest.json:
```diff
@@ -0,0 +1,17 @@
+    "name": "Maiana",
+    "version": "2024.07.0",
+    "home_assistant_domain": "esphome",
+    "funding_url": "https://github.com/marcelrv",
```
I think I would need to change that to point to the 'original' repo.

marcelrv replied:
See https://github.com/muebau/MAIANA-WiFi/pull/1#issuecomment-2304111067
Indeed, once you merge and setup a .io repository & page, you should indeed change these (unfortunately hardcoded) links to your repository

muebau commented on docu/manifest_firmware.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "Maiana",
```
MAIANA

muebau commented on platformio.ini:
I think this could use some cleanup and might need a change to fit with the idea of automatic builds. I don't know yet.

muebau commented on src/aisstore.cpp:
```diff
@@ -0,0 +1,95 @@
+
+#ifdef AISMEMORY
+
+#include <AIS.h>
```
Where does AIS.h come from? I guess GPSd oder https://github.com/KimBP/AIS? Did I missed it in platform.io?

muebau replied:
I can't compile the source as AIS.h is missing. What do I wrong there?

If `-D AISMEMORY` is disabled `ws_queue_client` is not defined.

```cpp
void onEvent(AsyncWebSocket *server, AsyncWebSocketClient *client,
             AwsEventType type, void *arg, uint8_t *data, size_t len) {
    switch (type) {
        case WS_EVT_CONNECT:
            // sendHistoryWS(client);
            ws_queue_client = client->id();
            ws_queue_msgId = 0;
            ws_queue_loop = millis();
            break;
```

marcelrv replied:
I indeed have it in my lib folder. from https://github.com/KimBP/AIS
I see it is somehow not in the commit, though also not in the .gitignore.

will check tonight. Prob better to add via the platformio.ini anyway

marcelrv replied:
I added [it](https://github.com/KimBP/AIS) via platformIO and directly recalled why that did not work....
It then tries to complile some example (I think for RPI) that gives loads of errors and prevents the thing from compiling.
Hence I copied the libarsry and removed the example and then it compiles fine without warnings or errors.

muebau commented on src/main.cpp:
```diff
@@ -5,90 +5,112 @@
-  Serial.println(X)
+#include "helperfunctions.h"
+
+#define CORS_HEADER  // For webdevelopment purpose. Adds CORS header
```
Shouldn't this deactivated for the production code?

marcelrv replied:
yes, I'll move it to the platformio.ini file and de-activate it

marcelrv replied:
I made it an app ooption now. (default disabled)

muebau commented on src/main.cpp:
```diff
@@ -180,857 +219,1155 @@ void forwardIt(const char *line);
+    server.on("/dashboard.html", HTTP_GET, [](AsyncWebServerRequest *request) {
+        request->send(SPIFFS, "/dashboard.html", "text/html");
+    });
+    //    server.serveStatic("/", SPIFFS, "/").setDefaultFile("index.html");
```
I thought it would be cool if there would be some fallback

marcelrv replied:
I was hesitant for that, as I can imagine at some point the config page you may want to password protect. 
If all files are readable the password would be easy redable too.

If all files would be readable directly, than in some ways it may be better to pull the config directly from the json files instead of having specific mout points for them..

muebau replied:
makes sense, good thoughts

muebau commented on src/main.cpp:
```diff
@@ -180,857 +219,1155 @@ void forwardIt(const char *line);
+    stopTCPNMEAForward();
+    if (mDNSOK) {
+        mdns_service_remove("_nmea-0183", "_tcp");
+        mdns_service_remove("_nmea-0183", "_udp");
```
Is it correct that the advertisement of the websocket is not stopped at this point?

marcelrv replied:
The mdns part of the code still needs bit of attention.
https://docs.espressif.com/projects/esp-protocols/mdns/docs/latest/en/index.html

I think also the ' addservice' part is wrong and needs to be fixed

muebau replied:
Yes I remember some differences between the ESP-IDF and the Arduino example. I found it again.

In `ESPmDNS.cpp` is the `addService(char *name, char *proto, uint16_t port)` which calls `mdns_service_add...`.

I think it could be changed to:
```cpp
mdns_service_add(NULL, "_http", "_tcp", 80, NULL, 0);
mdns_service_add(NULL, "_nmea-0183", "_tcp",  protocolSettings.tcpPort, NULL, 0);
```
etc. directly.

muebau replied:
At some point in the future there could be support for NMEA OneNet as soon as the final specs are released and it gets some adoption.

marcelrv replied:
Yes, but when I tried that, it did not like it as the arduino library used only allows for the `  bool addService(const char *service, const char *proto, uint16_t port)`
But also does not expose the feature that allows for multiple service names with the same protocol.
So for now I renamed the websocket variant `_nmea-0183-ws` to avoid the error.

Will add that commit once I tested some more and also added some of the other fixes

muebau commented on src/main.cpp:
```diff
@@ -180,857 +219,1155 @@ void forwardIt(const char *line);
+void forwardIt(const char *line) {
+    // if (networkOK && protocolSettings.port > 0 &&
+    //    protocolSettings.port < 65536) {
+    if (tcpForwaredOK) {
```
OK I think this typo is my fault, it should be 'tcpForwardOK' or so, right?

muebau commented on src/main.cpp:
```diff
@@ -180,857 +219,1155 @@ void forwardIt(const char *line);
-void safeProtocolToFile()
-{
+
+/* Does seem to get the websocket stuck. Remove
```
So this block could be removed right?

muebau reviewed (changes requested):
I love this contribution! 👍 I hope the comments help a bit to make give it the last finish.

marcelrv commented:
> I love this contribution! 👍 I hope the comments help a bit to make give it the last finish.

Thanks.
Fully agree to the comments.
Will do some more cleanup


Just to have your thought... do you think it is nice to have the ' demo mode' / test messages as an option you can switch-on/off in the UI. I was thinking about that, would be fairly easy as the sending is controlled by boolean. Would it be confusing or helping ?

muebau commented on data/dashboard.html:
```diff
@@ -0,0 +1,560 @@
+                        var speed = data.speed;
+
+                        textSpeed.text(speed);
+                        if (typeof speec === 'undefined' || speed === 'NaN') {
```
I think I found a small bug 'speed' instead of 'speec', right?

muebau reviewed (commented):
little bug?

muebau commented:
> Thanks. Fully agree to the comments. Will do some more cleanup

Well it doesn't mean that I am unwilling to do the work 😉. Some of the comments are more an expression of the idea I had while reading the code.

> 
> Just to have your thought... do you think it is nice to have the ' demo mode' / test messages as an option you can switch-on/off in the UI. I was thinking about that, would be fairly easy as the sending is controlled by boolean. Would it be confusing or helping ?

Well I love this feature. I used a less sophisticated version to test some functions. I think this could be very helpful. If it comes down (memory constraints) to this or the font I would personally rank it higher as the font.

marcelrv committed:
Configurable app settings for demo messages, CORS & AIS history

marcelrv commented:
> Well I love this feature. I used a less sophisticated version to test some functions. I think this could be very helpful. If it comes down (memory constraints) to this or the font I would personally rank it higher as the font.

Turned out it was not the memory constraints afterall. The SPIFF build error was the result of the path/filename of the font file being longer than 32 characters, which appears to be the max allowed for SPIFF

marcelrv committed:
improve reconnections for websocket. refactor CORS, MDNS fixes

marcelrv committed:
Automatic creation of merged files

marcelrv committed:
improve example webflasher

marcelrv commented:
@muebau I made several changes. Think i addressed the comments.

Was also doing some more experimentation on the webflasher. Did not manage to make the links relative. When using relative links it seems there is a CORS error.

Anyway, so far I found that it is not possible to directly flash from a regular github page. You need to publish it to the github webpages. The current html file I copy pasted to a github.io repository. See the result here: https://marcelrv.github.io/

Long story short, it works, you can flash the latest firmware with the current integration. (the python script will trigger a new firmware .bin file creation when you change the version in platformio.ini) But if the html page changes it needs to be manually deployed to the .io page. If that would happen frequent, you could make some publish script.

marcelrv committed:
resolve merge conflict for readme

Signed-off-by: marcel@verpaalen.com
Signed-off-by: Marcel Verpaalen <marcel@verpaalen.com>

marcelrv committed:
fix typos and improve readme

marcelrv committed:
improve readme

marcelrv committed:
minor readme improvements

marcelrv committed:
add mdns name to the readme

marcelrv committed:
include ais lib

muebau commented on src/main.cpp:
Perfect. It compiled and is running.

As I am not able to check with the AIS at this moment I tried "Demo Mode" and the ESP32 crashes reliably (guru meditation).

Do you see any reason why a Wemos ESP32 D1 mini (clone) should work? Do you use a ESP32 with more RAM?

```
Demo NMEA inputline: !AIVDM,1,1,,B,B3449BP0<84saC7HK=LDCwP5oP06,0*47
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400da897  PS      : 0x00060830  A0      : 0x800d667d  A1      : 0x3ffd3420  
A2      : 0x00000000  A3      : 0x15c7311a  A4      : 0x15c7311a  A5      : 0x3ffc4e40  
A6      : 0x00000001  A7      : 0x3ffd7ab0  A8      : 0x000000c9  A9      : 0x0000005a  
A10     : 0x3ffd85e8  A11     : 0x3ffd85e8  A12     : 0x000000c9  A13     : 0x00000001  
A14     : 0x3ffd5160  A15     : 0x0000005a  SAR     : 0x00000015  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000024  LBEG    : 0x40089638  LEND    : 0x4008964e  LCOUNT  : 0xffffffff  

```

muebau replied:
It seems like that the config page triggers the crash. I will investigate it deeper.

marcelrv replied:
I use very basic ESP32 from ali. Would not expect memory differences.

2 thoughts... 
1) I had Guru Meditation Error when I had the old web library included. Maybe cleaning the project helps to get rid of the previous one.
2) I have tested mostly with the protocol.json already on the filesystem. I'll  check if that maybe trigger the error if it is not included

marcelrv replied:
tried the 2nd, but could not trigger the error.

You could try once if it also happens when flashing from the test webpage, or upload the precompiled one.
If you have the same error, indeed it might be that there is a memory difference. If you don't have the error, we know it is somewhere a difference inthe build. Help narrow down the issue

muebau replied:
I don't know why but I think the protocol.json is not stored to the filesystem (see last few lines):

```
Webserver started
File saved successfully: /protocol.json
{"tcp":true,"tcpport":10110,"udp":false,"udpport":2000,"websocket":true,"websocketport":80}
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400da89b  PS      : 0x00060830  A0      : 0x800d6675  A1      : 0x3ffd4090  
A2      : 0x00000000  A3      : 0x3f40956d  A4      : 0x00000050  A5      : 0x00000001  
A6      : 0x00000027  A7      : 0x3ffd4724  A8      : 0x800e9190  A9      : 0x3ffd4090  
A10     : 0x3ffd416c  A11     : 0x00000000  A12     : 0x00000002  A13     : 0x0000ff00  
A14     : 0x00ff0000  A15     : 0xff000000  SAR     : 0x00000015  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000024  LBEG    : 0x4008aa49  LEND    : 0x4008aa59  LCOUNT  : 0xffffffff  


Backtrace: 0x400da898:0x3ffd4090 0x400d6672:0x3ffd40d0 0x400d6702:0x3ffd40f0 0x400d95f0:0x3ffd4110 0x400d9641:0x3ffd41a0 0x400e3ffd:0x3ffd41c0 0x400e0c0d:0x3ffd4210 0x400e18d1:0x3ffd4250 0x400e1a91:0x3ffd4290 0x400da5cd:0x3ffd42b0 0x400da65d:0x3ffd42f0 0x400daec2:0x3ffd4310




ELF file SHA256: ac579337c67bf53e

Rebooting...
<aHdf�]�bQsb<d␟`pb␟`df�␟`␅b␟�acba`pSd`�Maiana AIS Transponder forwarder started
SPIFFS ready
Protocol settings loaded from /protocol.json
File not found
Failed to load App settings. Using defaults
```

marcelrv replied:
the file not found is coming from the app.json file not being loaded.
I'll make the messages in the readjson and writejson contain the filename, than it is more clear.

muebau replied:
I tried to read the files from the filesystem but I was not successful. I think it crashes every time 'config.html' is load. My current best guess is that some config file does not exist or is not correct but is load by the code.

marcelrv replied:
Very odd.
The only thing I can think of is to read the files one by one to deeper understand what is causing the crash.

so, instead of loading config.html load
http://yourdevice/info
http://yourdevice/txstate
http://yourdevice/wifi
http://yourdevice/protocol
http://yourdevice/station
http://yourdevice/system
http://yourdevice/app
and see which one causes the crash

marcelrv replied:
Did you have the chance to try to test / load the precompiled version to see if the same happens?

marcelrv replied:
The only other thought I can imagine: version differences in the framework:
![image](https://github.com/user-attachments/assets/a1fb3140-0971-4342-b388-6dbb29c6b883)
I'll update to the latest, see if that makes any difference

muebau replied:
The problem seems to be the app.json. I see that the app.json is not written. The debug output from the 'writeJsonFile' doesn't show up.

'Uncaught (in promise) TypeError: NetworkError when attempting to fetch resource.'

```javascript
        function refresh_info() {
            if (document.getElementById('station-mmsi').value === '') {
                console.log('station data empty.. refreshing');
                refresh_station();
            }
        }

```

muebau replied:
> Did you have the chance to try to test / load the precompiled version to see if the same happens?

I didn't had a chance to flash it yet

muebau replied:
> The only other thought I can imagine: version differences in the framework: ![image](https://private-user-images.githubusercontent.com/5871213/364783116-a1fb3140-0971-4342-b388-6dbb29c6b883.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1NDE0NjIsIm5iZiI6MTcyNTU0MTE2MiwicGF0aCI6Ii81ODcxMjEzLzM2NDc4MzExNi1hMWZiMzE0MC0wOTcxLTQzNDItYjM4OC02ZGJiMjljNmI4ODMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDVUMTI1OTIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODM3NjgxZGY5ZGJiOTk2MmViYTQ3M2Q4MzA3Njk1OWMyMzEwOWY2ZjYzN2U1M2I0MjU2YzIwMWU2MTVkYWYwZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.3t6sUfiRoyc1wOVAjL5eVX2HFh6Pv88cLU4MMCOeAoc) I'll update to the latest, see if that makes any difference

good idea. I will check that too

muebau replied:
> Very odd. The only thing I can think of is to read the files one by one to deeper understand what is causing the crash.
> 
> so, instead of loading config.html load http://yourdevice/info http://yourdevice/txstate http://yourdevice/wifi http://yourdevice/protocol http://yourdevice/station http://yourdevice/system http://yourdevice/app and see which one causes the crash

that is the next thing I test

marcelrv replied:
I did a full clean. these are the versions
 ```
Processing wemos_d1_mini32 (platform: espressif32; board: wemos_d1_mini32; framework: arduino)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------Tool Manager: Installing platformio/framework-arduinoespressif32 @ ~3.20017.0
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Tool Manager: framework-arduinoespressif32@3.20017.0 has been installed!
Library Manager: Installing bblanchon/ArduinoJson @ ^6.19.1
Unpacking  [####################################]  100%
Library Manager: ArduinoJson@6.21.5 has been installed!
Library Manager: Installing mathieucarbou/AsyncTCP @ ^3.2.4
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: AsyncTCP@3.2.5 has been installed!
Library Manager: Installing mathieucarbou/ESPAsyncWebServer @ 3.1.5
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: ESPAsyncWebServer@3.1.5 has been installed!
Library Manager: Resolving dependencies...
Library Manager: Installing ayushsharma82/ElegantOTA @ ^3.1.1
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: ElegantOTA@3.1.5 has been installed!
Library Manager: Resolving dependencies...
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/wemos_d1_mini32.html
PLATFORM: Espressif 32 (6.8.1) > WEMOS D1 MINI ESP32
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-arduinoespressif32 @ 3.20017.0 (2.0.17)
 - tool-esptoolpy @ 1.40501.0 (4.5.1)
 - tool-mkfatfs @ 2.0.1
 - tool-mklittlefs @ 1.203.210628 (2.3)
 - tool-mkspiffs @ 2.230.0 (2.30)
 - tool-openocd-esp32 @ 2.1100.20220706 (11.0)
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 38 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ArduinoJson @ 6.21.5
|-- AsyncTCP @ 3.2.5
|-- ESPAsyncWebServer @ 3.1.5
|-- ElegantOTA @ 3.1.5
|-- AIS @ 0.1.0
|-- SPIFFS @ 2.0.0
|-- FS @ 2.0.0
|-- ESP32 Async UDP @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- WiFi @ 2.0.0
```

muebau replied:
I cleaned everything and did the best to update everything to the newest versions. With an ESP32 Wemos D1 mini clone and an ESP32 Node MCU:

* erased flash
* full clean project
* clean
* upload filesystem image
* upload & monitor
* connect to WiFi
* browse to 'maiana.local'
* click on 'config'
* **Guru Mediation**

Are your ESP32 get hot too?
```
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/wemos_d1_mini32.html
PLATFORM: Espressif 32 (6.8.1) > WEMOS D1 MINI ESP32
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20017.0 (2.0.17) 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-mkfatfs @ 2.0.1 
 - tool-mklittlefs @ 1.203.210628 (2.3) 
 - tool-mkspiffs @ 2.230.0 (2.30) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 38 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ArduinoJson @ 7.1.0
|-- AsyncTCP @ 3.2.5
|-- ESPAsyncWebServer @ 3.2.4
|-- ElegantOTA @ 3.1.5
|-- AIS @ 0.0.0+20240730234923
|-- SPIFFS @ 2.0.0
|-- FS @ 2.0.0
|-- ESP32 Async UDP @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- WiFi @ 2.0.0
Building in release mode

```

marcelrv replied:
> Are your ESP32 get hot too?

no, warmer than outside. Would guess 30-35 degrees

muebau replied:
>  upload the precompiled one.

I did that and it crashes on both ESP32 I have here. The question is why?
I am sorry for all the trouble.

marcelrv replied:
I'll try if I can find a ESP32 Wemos D1 mini, I might have one somewhere here at home.
See if I can somehow replicate the matter with that one.

muebau replied:
I did
* removed the PlatformIO framework completely
* got a fresh copy of the code from Github
* erased the flash of the D1 mini
* wrote a new filesystem & firmware

What works
* http://maiana.local/
* http://yourdevice/info
* http://yourdevice/txstate
* http://yourdevice/wifi
* http://yourdevice/protocol
* http://yourdevice/station
* http://yourdevice/system
* http://yourdevice/app
* http://maiana.local/config.html
* http://maiana.local/dashboard.html
* http://maiana.local/app?demoMode=false&corsHeader=true&aisMemory=true
* http://maiana.local/station?mmsi=&callsign=&vesselname=&vesseltype=&loa=0&beam=0&portoffset=0&bowoffset=0
* http://maiana.local/wifi?type=ap&ssid=MAIANA&password=MAIANA-AIS

**What crashes (Guru)**
* http://maiana.local/protocol?tcp=false&tcpport=10110&udp=false&udpport=2000&websocket=true&websocketport=80

The ESP32 gets so hot that it is not comfortable to touch.
Idea:
I have nothing connected to the ESP32 D1 mini so there flies a lot of garbage text through the terminal (it may contribute to the load) and the config button may also flicker.

The interesting question is now why the call of `http://maiana.local/config.html` works without a crash and why only `http://maiana.local/protocol?...` is left.

marcelrv replied:
I did bit of googling on the ESP32 Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled. Most seem to point to doing too much while it was in an interrupt, OR memory errors OR NPE's

For the ' doing too much during interrupt' in the very last commit I did I moved the start & stopNMEA forward only to be done while writing changing the parameters, not just when requesting the parameters.
==> so when you did your request only with ? then the save and start/stop were not executed

https://github.com/muebau/MAIANA-WiFi/pull/1/files#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2R652-R656

One way to maybe avoid that is to move the saving of the file outside of the response and inside the loop.
(somewhat similar what I did with the websocket data sending). (like adding a bool protocolSettingChanged. and then in the loop have a if protocolSettingChanged then save and NMEA stop and start).

marcelrv committed:
rename webservice mdns

marcelrv committed:
verbose file operations and app.json writing

marcelrv committed:
minor improvements

marcelrv commented:
Found one... Will try today
![IMG_20240906_123239](https://github.com/user-attachments/assets/74180169-e20d-4473-91d1-ca882b3887fa)

marcelrv commented:
"good news" I'm able to replicate the guru error.  meaning on a fresh install I have the same error.

my findings so far: 
* It really gets warm when the device is in AP mode or STA+AP
* I can load all the individual requests, but at the loading on the config page, when all request come at the same time it gives GURU error
* Once wifi is defined and connected to normal network I can no longer replicate the issue. I can do the loading of the page, or use the refresh button without it going into the guru error
* When I change back to AP mode it seems to work as well, without any guru errors.

marcelrv committed:
Avoid GURU error during initial setup

marcelrv commented:
I moved the serial.writes outside the webrequests. After that I could not replicate the issue anymore.
Maybe you can give it a try

muebau commented:
> I moved the serial.writes outside the webrequests. After that I could not replicate the issue anymore. Maybe you can give it a try

So 'cool' you was able to reproduce the error. On the other hand I am glad that it was not bad hardware on my site. I started to worry about that.

It is right I used the AP mode only. But I think this mode would be the preferred for most of the users here in the harbor (it started to get some attention) 😉. Most have a existing marine radio with an inbuilt AIS receiver and would like to use the ESP32 in AP mode.

muebau commented:
I guess I solved the misery!!!
1. I just realized that it might be a good idea to avoid to call things on a nullptr
2. I should try to avoid a memory leak by releaving the uniqe poiinter from its duties but forget to finish the task then.

So after some changes the code went from

```
void stopTCPNMEAForward()
{
    tcpForwardOK = false;
    tcp.get()->end();
    tcp.release();
}
```
to

```
void stopTCPNMEAForward()
{
    if (tcp)
    {
        tcp.reset();
    }
}
```

muebau commented:
I need to revert all the other things I tried but I think that was the culprit.

muebau commented:
> ```
> void stopTCPNMEAForward()
> {
>     if (tcp)
>     {
>         tcp.reset();
>     }
> }
> ```

That solved it indeed. I am not able to trigger any Guru meditation anymore. 🎉

Should I then merge and create a release?

marcelrv commented:
Would be good. 
Now there are no guru errors anymore, it is a good new basis.
